### DISCUSSION TITLE
Is it possible for OpenThread to run on top of the Linux kernel's IEEE 802.15.4 subsystem?

### DISCUSSION THREAD

**Discussion by objelf** [openthread/openthread#10149] [Q&A]
Hello everyone,

I'm new to the OpenThread stack. Recently, I've been interested in writing an IEEE 802.15.4 driver for the official Linux kernel.
This would allow me to take advantage of existing features such as TCP/IP, 6LoWPAN, and the software MAC feature and so on already in the Linux kernel. The subsystem I'm referring to is located here in the Linux kernel tree: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ieee802154

I'm wondering if OpenThread can run on top of the Linux kernel's IEEE 802.15.4 subsystem, or if it runs completely independently from the Linux kernel, directly communicating with the 802.15.4 hardware? 

Thank you to everyone who shared ideas on how I can write a driver for new hardware for OpenThread using the Linux kernel.

**Comment by jwhui** [comment]
OpenThread does not readily integrate with the Linux IEEE 802.15.4 subsystem. OpenThread directly integrates with an IEEE 802.15.4 radio. OpenThread also presents a standard Linux network interface (using tun), allowing direct integration into the Linux IP stack.

**Reply by objelf** [reply]
Hi jwhui, thank you for getting back to me so quickly. I'm new to the Thread protocol and OpenThread stack, so I have another question. Does OpenThread implement the IEEE 802.15.4 protocol? I meant does it encapsulate and decapsulate packets before sending or receiving to/from IEEE 802.15.4 radio and handle the IEEE 802.15.4 softMAC? The hardware I'm working with functions as a pure IEEE 802.15.4 PHY.

**Reply by abtink** [reply]
OpenThread implement 15.4 sub-mac and mac layers. 
It defines a set of `otPlatRadio` APIs  that the radio platform layer should provide.
Please check out the [`openthread/platform/radio.h`](https://github.com/openthread/openthread/blob/main/include/openthread/platform/radio.h).

**Reply by objelf** [reply]
Hi Abtink, I appreciate your assistance.
Does somebody know if there's a vendor driver available for setting up a scenario where the driver controls the 802.15.4 PHY using SPI or UART, handling firmware setup and initialization? I'm looking for a reference to take the first step. Thanks for your help in getting me started with Openthread!

**Reply by jwhui** [reply]
There are a number of radio SoC vendors that support OpenThread: https://openthread.io/platforms
